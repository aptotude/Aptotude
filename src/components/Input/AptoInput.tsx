import React from 'react';
import classNames from 'classnames';
import './aptoInput.scss';
import {
  AptoFormControl,
  AptoFormControlDisplayProps
} from '../FormControl/AptoFormControl';

const COMPONENT_PREFIX = 'AptoInput';

export interface AptoInputDisplayProps {
  className?: string;
  inputClasses?: string;
  defaultValue?: any;
  disabled?: boolean;
  error?: React.ReactNode;
  id?: string;
  label?: React.ReactNode;
  maxLength?: number;
  multiline?: boolean;
  name?: string;
  onBlur?: React.FormEventHandler<HTMLTextAreaElement | HTMLInputElement>;
  onChange?: React.ChangeEventHandler<HTMLTextAreaElement | HTMLInputElement>;
  onClick?: React.MouseEventHandler<HTMLTextAreaElement | HTMLInputElement>;
  onFocus?: React.FocusEventHandler<HTMLTextAreaElement | HTMLInputElement>;
  onKeyPress?: React.KeyboardEventHandler<
    HTMLTextAreaElement | HTMLInputElement
  >;
  placeholder?: string;
  readOnly?: boolean;
  required?: boolean;
  rows?: number;
  type?: string;
  value?: any;
  inline?: boolean;
}

export class AptoInput extends React.Component<AptoInputDisplayProps> {
  public static defaultProps = {
    multiline: false,
    required: false,
    rows: 1,
    type: 'text'
  };

  public inputNode: HTMLTextAreaElement | HTMLInputElement | null = null;

  public handleChange = (
    event: React.ChangeEvent<HTMLTextAreaElement | HTMLInputElement>
  ) => {
    const { onChange } = this.props;

    if (onChange) {
      onChange(event);
    }
  };

  public focus() {
    this.inputNode!.focus();
  }

  public blur() {
    this.inputNode!.blur();
  }

  public setNode = (node: HTMLTextAreaElement | HTMLInputElement | null) => {
    this.inputNode = node;
  };

  public render() {
    const {
      className,
      defaultValue,
      disabled,
      error,
      label,
      multiline,
      onChange,
      readOnly,
      required,
      rows,
      type,
      inline,
      inputClasses,
      id,
      ...rest
    } = this.props;

    const InputComponent = multiline ? 'textarea' : 'input';

    const classes = classNames(
      COMPONENT_PREFIX,
      `${COMPONENT_PREFIX}--${multiline ? 'textarea' : 'input'}`,
      error && `${COMPONENT_PREFIX}--error`,
      inputClasses
    );

    const formGroupProps: AptoFormControlDisplayProps = {
      className,
      label,
      error,
      required,
      readOnly,
      disabled,
      inline,
      id
    };

    return (
      <AptoFormControl {...formGroupProps}>
        <InputComponent
          className={classes}
          readOnly={readOnly}
          required={required}
          disabled={disabled}
          type={InputComponent === 'input' ? type : undefined}
          rows={InputComponent === 'textarea' ? rows : undefined}
          onChange={this.handleChange}
          ref={this.setNode}
          {...rest}
        />
      </AptoFormControl>
    );
  }
}
