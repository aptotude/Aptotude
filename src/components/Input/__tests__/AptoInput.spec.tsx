import * as React from 'react';
import { render, cleanup } from 'react-testing-library';
import { AptoInput } from '../AptoInput';

afterEach(() => {
  cleanup();
});

describe('AptoInput', () => {
  it('Renders Component', () => {
    const { container } = render(<AptoInput />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('Renders with name/value', () => {
    const { container } = render(<AptoInput name="foo" value="bar" />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('Renders disabled', () => {
    const { container } = render(
      <AptoInput disabled={true} name="foo" value="bar" />
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
