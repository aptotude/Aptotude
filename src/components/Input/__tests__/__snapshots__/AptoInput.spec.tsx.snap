// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AptoInput Renders Component 1`] = `
<div
  class="ui input"
>
  <input
    type="text"
  />
</div>
`;

exports[`AptoInput Renders disabled 1`] = `
<div
  class="ui disabled input"
>
  <input
    disabled=""
    name="foo"
    tabindex="-1"
    type="text"
    value="bar"
  />
</div>
`;

exports[`AptoInput Renders with name/value 1`] = `
<div
  class="ui input"
>
  <input
    name="foo"
    type="text"
    value="bar"
  />
</div>
`;
